pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -0,0 +1,219 @@
+			c, err := a.oauthClient.Get(name, metav1.GetOptions{})
+			if err != nil {
+				if kerrors.IsNotFound(err) {
+					delete(a.clientTimeouts, name)
+				}
+			}
+			if c.AccessTokenTimeoutSeconds == nil {
+				delete(a.clientTimeouts, name)
+			} else {
+				a.clientTimeouts[name] = *c.AccessTokenTimeoutSeconds
+				// Use half period if it is smaller
+				half := int32(math.Ceil(float64(a.clientTimeouts[name]) / 2))
+				if flush > half {
+					flush = half
+				}
+			}
+		}
+	}
+	return time.Duration(flush) * time.Second
+}
+
+func (a *TokenTimeoutUpdater) clientTimeout(name string) int32 {
+	t, ok := a.clientTimeouts[name]
+	if !ok {
+		c, err := a.oauthClient.Get(name, metav1.GetOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				glog.Info("Token timeout set but OAuthClient %s not found!", name)
+			} else {
+				glog.Error("Failed to fetch OAuthClient for timeout value")
+			}
+			return 0
+		}
+		if c.AccessTokenTimeoutSeconds == nil {
+			glog.Info("Token timeout set but OAuthClient %s has no timeout!", name)
+			return 0
+		}
+		t = *c.AccessTokenTimeoutSeconds
+		a.clientTimeouts[name] = t
+	}
+	return t
+}
+
+func (a *TokenTimeoutUpdater) insert(td tokenData) {
+	a.data[td.name] = td
+	a.tree.ReplaceOrInsert(&tokenDataRef{td.name, td.timeout})
+}
+
+func (a *TokenTimeoutUpdater) remove(td tokenData) {
+	a.tree.Delete(&tokenDataRef{td.name, td.timeout})
+	delete(a.data, td.name)
+}
+
+func (a *TokenTimeoutUpdater) flush(flushHorizon time.Time) {
+	flushedTokens := 0
+	totalTokens := len(a.data)
+
+	glog.Infof("Flushing tokens timing out before %v", flushHorizon)
+
+	for {
+		item := a.tree.Min()
+		if item == nil {
+			// out of items
+			break
+		}
+		td := a.data[item.(*tokenDataRef).name]
+		if td.timeout.Before(flushHorizon) {
+			delta := a.clientTimeout(td.client)
+			// ignore tokens for which we can't find a delta
+			if delta > 0 {
+				token, err := a.tokens.Get(td.name, metav1.GetOptions{})
+				if err == nil { // just ignore if the token is gone
+					newtimeout := int32(td.seen.Sub(token.CreationTimestamp.Time)/time.Second) + delta
+					if newtimeout > token.TimeoutsIn {
+						token.TimeoutsIn = newtimeout
+						// ignore errors, if it fails to update there isn't much we can do
+						_, _ = a.tokens.Update(token)
+					}
+				}
+			}
+			a.remove(td)
+			flushedTokens++
+		} else {
+			// out of items within the flush Horizon
+			break
+		}
+	}
+
+	glog.Infof("Flushed %d tokens out of %d in bucket", flushedTokens, totalTokens)
+}
+
+func (a *TokenTimeoutUpdater) Start() {
+	// TODO: make safety margin configurable ?
+	var safetyMargin time.Duration = time.Second * 15
+
+	glog.Infof("Started Token Timeout Flush Handling thread!")
+
+	flushTimeout := a.updateClientTimeouts()
+	nextTimer := time.NewTimer(flushTimeout)
+	nextTimeout := time.Now().Add(flushTimeout)
+
+	for {
+		select {
+		case td, ok := <-a.tokenChannel:
+			if !ok {
+				// when channel closes terminate
```

Comment:
You will want the stop channel from the post start hook to trigger this.

Reply:
I looked around, it seem origin everywhere passes in wait.NeverStops() .. so I did not see the point anyway... If I can figure out how to get started in the post start hook then I might do it anyways

Reply:
Kube uses it correctly - openshift may be lagging behind since David started working on getting that stuff sorted in 3.7